### ISSUE TITLE
Apps with dependency of Connectivity package require NSLocationAlwaysAndWhenInUseUsageDescription

### ISSUE THREAD

**Issue by Vanethos** [flutter/flutter#40418]
After using this plugin for a couple of months with no problem, Apple sent me the following e-mail yesterday after sending another build:

```
    Dear Developer, We identified one or more issues with a recent delivery for your app, "Ianum" 1.1.2 (131).
    Your delivery was successful, but you may wish to correct the following issues in your next delivery: I

    TMS-90683: Missing Purpose String in Info.plist - Your app's code references one or more APIs that access sensitive user data. The app's Info.plist file should contain a NSLocationAlwaysUsageDescription key with a user-facing purpose string explaining clearly and completely why your app needs the data. Starting Spring 2019, all apps submitted to the App Store that access user data are required to include a purpose string. If you're using external libraries or SDKs, they may reference APIs that require a purpose string. While your app might not use these APIs, a purpose string is still required. You can contact the developer of the library or SDK and request they release a version of their code that doesn't contain the APIs. Learn more (https://developer.apple.com/documentation/uikit/core_app/protecting_the_user_s_privacy).

    ITMS-90683: Missing Purpose String in Info.plist - Your app's code references one or more APIs that access sensitive user data. The app's Info.plist file should contain a NSLocationWhenInUseUsageDescription key with a user-facing purpose string explaining clearly and completely why your app needs the data. Starting Spring 2019, all apps submitted to the App Store that access user data are required to include a purpose string. If you're using external libraries or SDKs, they may reference APIs that require a purpose string. While your app might not use these APIs, a purpose string is still required. You can contact the developer of the library or SDK and request they release a version of their code that doesn't contain the APIs. Learn more (https://developer.apple.com/documentation/uikit/core_app/protecting_the_user_s_privacy).

    After you’ve corrected the issues, you can use Xcode or Application Loader to upload a new binary to App Store Connect.
    Best regards, The App Store Team
```

Our app **does not** use the functionality of retrieving IP address, WIFI or BSSID name, and as such, I'd like to ask you the following:
* When adding strings to the `info.plist` will the user be able to see that the app accesses the location? Be it in the App Store, or the OS? I ask this because as a former Android Developer I remember that everything that we put on the manifest the user could see.
* If that is the case, what can we do to **not** have those strings in the info.plist file?
* And finally, why is this only flagged now after using the plugin for at least 2 months with no upgrades in the meantime? Coincidentally, this happened after I upgraded to 1.9.

As a solution for this, if no other option is available, I'll create a local fork of the package that only contains the connectivity part and use it for this project.

Our app is privacy-focused and we want to be fully transparent with our users, so having permissions that we are not using without explaining it is not an option.

**Comment by Vanethos** [comment]
In the meantime, I used the code of the plugin to create a modified package that does not have any location-aware code. You can find it here: https://github.com/Vanethos/simple_connectivity

I didn't make a PR since people might want to know the wifi information so maybe this plugin should be divided into two.

Nonetheless, if you don't want it to be divided into two, I ask you if it is possible for me to publish this package so that I can use it in my projects and plugins that I have created.

**Comment by cyanglaz** [comment]
@Vanethos 

To answer your questions:

> * When adding strings to the `info.plist` will the user be able to see that the app accesses the location? Be it in the App Store, or the OS? I ask this because as a former Android Developer I remember that everything that we put on the manifest the user could see.

Once you added connectivity plugin with adding strings in `info.plist`, user will see your app might request location permission in the home page of your app in the app store. Also, if you make the location related method calls, they will see the string you put in the `info.plist` in a pop-up alert view..

> * If that is the case, what can we do to **not** have those strings in the info.plist file?

I don't have any knowledge on anyway to avoid this. It is required by iOS.

> * And finally, why is this only flagged now after using the plugin for at least 2 months with no upgrades in the meantime? Coincidentally, this happened after I upgraded to 1.9.

Starting from iOS 13. Apple requires apps using certain functionality to obtain the location permissions from users. We added the location related code to handle this new requests.
You can find details here: https://github.com/flutter/flutter/issues/35332

If you concern about showing your app requesting location permission in the home page of your App, I would say your current work around is a good way to go.

**Comment by ccadieux** [comment]
@Vanethos Thanks for the fork. I have this issue as well. 

I really only need to know the state of the network. We want to give the user the option to turn off uploads while on cell. We use none of the features that would require location. I do think the package should be split up.

It would be nice if the package stayed true to the description.
```This plugin allows Flutter apps to discover network connectivity and configure themselves accordingly. It can distinguish between cellular vs WiFi connection. This plugin works for iOS and Android.```

**Comment by CoreyCole** [comment]
This is a similar problem that [flutter permission handler](https://github.com/BaseflowIT/flutter-permission-handler/issues/26#issuecomment-541974767) has. Since it allows you to request permission for all, you have to declare how you are using all in your `info.plist`. My workaround was to declare that we are not using those permissions in the description:
```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	// ...
	<key>NSAppleMusicUsageDescription</key>
	<string>This app does not access your music.</string>
	<key>NSCalendarsUsageDescription</key>
	<string>This app does not access your calendars.</string>
	<key>NSContactsUsageDescription</key>
	<string>This app does not access your contacts.</string>
	<key>NSLocationAlwaysUsageDescription</key>
	<string>This app does not access your location.</string>
	<key>NSLocationWhenInUseUsageDescription</key>
	<string>This app does not access your location.</string>
	<key>NSMicrophoneUsageDescription</key>
	<string>This app uses the microphone to record audio.</string>
	<key>NSMotionUsageDescription</key>
	<string>This app does not access motion.</string>
	<key>NSSpeechRecognitionUsageDescription</key>
	<string>This app does not access speech regonition.</string>
	// ...
</dict>
</plist>
```
But, this is probably not the best workaround for privacy-focused apps as it would be better if those weren't visible at all because you aren't using them.

Is this a hard limit of flutter/dart? Is the only way to solve this to break out different permissions into different packages?

IMO a good developer experience would be if these related features were in 1 package, but we only need to add the permission for what we are using. What we are using could be declarative by our imports, i.e.:
```dart
// we need to add Core Location permission to the info.plist
// because `connectivity.dart` exposes all the methods
import 'package:connectivity/connectivity.dart'; 
```
vs
```dart
// we don't need to add Core Location permission to the info.plist
// because `on_connectivity_changed.dart` doesn't expose WiFi info methods
import 'package:connectivity/on_connectivity_changed.dart'; 
```
But, I'm not sure if this is possible 🤔I welcome any input from those more experienced with dart/flutter platform channel plugins. Just trying to think outside the box here.

**Comment by cyanglaz** [comment]
Currently, we don't have a first party plugin to get location information. So the location related methods have to be in the plugin.
@Vanethos Are you ok if we close this issue for now? If you'd like a separate first party plugin that handles location so that `connectivity` can be free from any location related code, we can open another issue to track that effort.

**Comment by Vanethos** [comment]
So the job of connectivity is to handle wifi connection state plus GPS?

If that's the case, yes, it would be best to decouple both!

If needed I can work on that

**Comment by finpro-svn** [comment]
I would also welcome a separate first party plugin that handles location so that connectivity without any location related code. It is bad practice to request permissions that an app does not need just because the plugin requires them.

**Comment by CoreyCole** [comment]
@finpro-svn to be clear, the app is not requesting permissions. It is simply stating on the app store listing that the app might request them and here is the developer's rational for having them. My teams workaround was just to state that we are not using them and won't request them in the app store listing via the Info.plist
```
<key>NSLocationAlwaysUsageDescription</key>
<string>This app does not access your location and will not request location permissions.</string>
``` 
This might not be good enough for privacy focused apps, but it good enough for our use case. We never request the users location permission.

**Comment by jerryzhoujw** [comment]
I think separate or by someway not enable GPS location code would be reasonable.  This plugin's main functions should only about check connectivity.

**Comment by cyanglaz** [comment]
I agree with @jerryzhoujw, the plugin should separate from location services. We are currently working on looking for a good plugin in pub for that functionality and we will drop all the location based code in this plugin. @csells https://pub.dev/packages/location and https://pub.dev/packages/location_permissions look good.

**Comment by CoreyCole** [comment]
The permissions handler plugin was able to solve this without branching into different plugins https://github.com/Baseflow/flutter-permission-handler/issues/26#issuecomment-573298064

It looks like they do some pod script:
> You can remove permissions you don't use by:
> 1. Add the following to your `Podfile` file:
>
>    ```ruby
>    post_install do |installer|
>      installer.pods_project.targets.each do |target|
>        target.build_configurations.each do |config|
>          ... # Here are some configurations automatically generated by flutter
>    
>          # You can remove unused permissions here
>          # for more infomation: https://github.com/BaseflowIT/flutter-permission-handler/blob/develop/ios/Classes/PermissionHandlerEnums.h
>          # e.g. when you don't need camera permission, just add 'PERMISSION_CAMERA=0'
>          config.build_settings['GCC_PREPROCESSOR_DEFINITIONS'] ||= [
>            '$(inherited)',
>   
>            ## dart: PermissionGroup.calendar
>            # 'PERMISSION_EVENTS=0',
>   
>            ## dart: PermissionGroup.reminders
>            # 'PERMISSION_REMINDERS=0',
>   
>            ## dart: PermissionGroup.contacts
>            # 'PERMISSION_CONTACTS=0',
>   
>            ## dart: PermissionGroup.camera
>            # 'PERMISSION_CAMERA=0',
>   
>            ## dart: PermissionGroup.microphone
>            # 'PERMISSION_MICROPHONE=0',
>   
>            ## dart: PermissionGroup.speech
>            # 'PERMISSION_SPEECH_RECOGNIZER=0',
>   
>            ## dart: PermissionGroup.photos
>            # 'PERMISSION_PHOTOS=0',
>   
>            ## dart: [PermissionGroup.location, PermissionGroup.locationAlways, PermissionGroup.locationWhenInUse]
>            # 'PERMISSION_LOCATION=0',
>           
>            ## dart: PermissionGroup.notification
>            # 'PERMISSION_NOTIFICATIONS=0',
>   
>            ## dart: PermissionGroup.mediaLibrary
>            # 'PERMISSION_MEDIA_LIBRARY=0',
>   
>            ## dart: PermissionGroup.sensors
>            # 'PERMISSION_SENSORS=0'
>          ]
>   
>        end
>      end
>    end
>    ```
>
> 2. Delete the corresponding permission description in `Info.plist`
>
>    e.g. when you don't need camera permission, just delete 'NSCameraUsageDescription'
>
>    The following lists the relationship between `Permission` and `The key of Info.plist`:
>
>    | Permission  | Info.plist  | Macro |
>    |---|---|---|
>    | PermissionGroup.calendar  |  NSCalendarsUsageDescription | PERMISSION_EVENTS |
>    | PermissionGroup.reminders  |  NSRemindersUsageDescription | PERMISSION_REMINDERS |
>    | PermissionGroup.contacts  |  NSContactsUsageDescription | PERMISSION_CONTACTS |
>    | PermissionGroup.camera  |  NSCameraUsageDescription | PERMISSION_CAMERA |
>    | PermissionGroup.microphone  |  NSMicrophoneUsageDescription | PERMISSION_MICROPHONE |
>    | PermissionGroup.speech  |  NSSpeechRecognitionUsageDescription | PERMISSION_SPEECH_RECOGNIZER |
>    | PermissionGroup.photos  |  NSPhotoLibraryUsageDescription | PERMISSION_PHOTOS |
>    | PermissionGroup.location, PermissionGroup.locationAlways, PermissionGroup.locationWhenInUse  |  NSLocationUsageDescription, NSLocationAlwaysAndWhenInUseUsageDescription, NSLocationWhenInUseUsageDescription | PERMISSION_LOCATION |
>    | PermissionGroup.notification  |  PermissionGroupNotification | PERMISSION_NOTIFICATIONS |
>    | PermissionGroup.mediaLibrary  |  NSAppleMusicUsageDescription, kTCCServiceMediaLibrary | PERMISSION_MEDIA_LIBRARY |
>    | PermissionGroup.sensors  |  NSMotionUsageDescription | PERMISSION_SENSORS |
>
> 3. Clean & Rebuild

You can see the full PR here: https://github.com/Baseflow/flutter-permission-handler/commit/a1b96e6979bd29b79fd53be9e261af1666737bf1

**Comment by cyanglaz** [comment]
Un-assign myself from this issue as I don't have cycle to actively working on it.

**Comment by YAPPDEF** [comment]
Can you please provide a similar/the same solution as for the permissions handler plugin?

**Comment by KevinTheGray** [comment]
I opened a PR to handle this issue here: https://github.com/flutter/plugins/pull/2771

It uses a similar macro solution that `permission_handler` uses.

**Comment by cyanglaz** [comment]
Looks like we have 2 solutions:
1. Splitting the plugin into 2 plugins.

-  The connectivity plugin only contains `checkConnectivity`. 
-  A new `network_info` plugin contains wifi related APIs.
This is going to be a breaking change thus a major version bump is required. For people who needs to use the WIFI related APIs, they'd have to add a new dependency to the `network_info` plugin, and do some minor and straightforward code migrations. (This affects all the apps that uses connectivity plugin and all the packages depends on the connectivity plugin)

2. Conditionally add WIFI related code (flutter/plugins#2771). This is a backward compatible fix. We still keep the wifi related code in the code base, which might leads to other related issues in the future.

**Comment by KevinTheGray** [comment]
@cyanglaz what's the preferred path of the Flutter team?  I'd be ok with either approach, but it seems like the splitting of the packages would be the best path forward, since as you said it could cause related issues in the future.  It also reduces the amount of lift required by users of the package, as they would not need to modify the `podfile` for the desired behavior.

**Comment by brunobowden** [comment]
@cyanglaz @KevinTheGray - I'd appreciate raising the priority on this. This is a launch blocking issue for the official COVID-19 app of the World Health Organization that I'm helping to develop: https://github.com/WorldHealthOrganization/app/issues/1520

The WHO COVID-19 app is a Flutter based app developed by a large group of volunteers, including many Xooglers. The app doesn't use location for reasons of privacy and to make it possible to use the Apple / Google Exposure Notification API. Use of that API prohibits use of location permissions by policy. We would be happy with any solution but ideally something possible over the next week. I appreciate that the reason the issue has been around for a year is that it's far from a trivial issue to fix. I hope this can give it renewed impetus for a solution.

@Vanethos - would you mind updating the `simple_connectivity` plugin in the interim with the latest upstream version e.g. `0.4.9+3`? I was going to suggest using exactly the same version numbers as the `connectivity` plugin, so people could tell what version it was based on. The current versions listed of 0.1.0 and 0.1.1 made it hard to tell what it was based on. This would be helpful as an interim workaround... I'd just much prefer to use a more recent version. If you could do this, then I'd be much obliged.

**Comment by Vanethos** [comment]
Will try to do that today @brunobowden !

**Comment by Vanethos** [comment]
@brunobowden Just to be sure, if there are missing functionalities on the `simple_connectivity` please file an issue in the repo since it isn't as simple as syncing with the files from the upstream. We have to change the code in order to remove the dependencies.

**Comment by brunobowden** [comment]
@Vanethos - Moving simple_connectivity discussion to https://github.com/Vanethos/simple_connectivity/issues/7

@cyanglaz @KevinTheGray - This is definitely a tricky one.... but I think it's best resolved before a v1.0 release. If you don't resolve it beforehand, the dependency on location permissions is baked in to the v1.0 API and retains the guarantee of backwards compatibility. It's obviously awkward as the README mention of v1.0 implies some stability.... but from https://semver.org/#spec-item-4 for v0.y.z; _Anything MAY change at any time. The public API SHOULD NOT be considered stable._

I agree that it needs to be two different libraries but we could also use @KevinTheGray's approach to ease the transition on both sides. I would defer to @cyanglaz's judgement but here's my suggestion.

1) v0.5: Review and merge in https://github.com/flutter/plugins/pull/2771. Bump to v0.5 and add warning to README about likely breaking change.
2) I and others start experimenting with `DISABLE_CONNECTIVITY_LOCATION_CODE` Podfile workaround to confirm that this works as intended. Existing devs have no disruption.
3) @cyanglaz and/or others have time to split out `network_info` and get that operational. The README will then mention the steps needed for forward compatibility by transitioning to `network_info`.
4) v0.6: Remove `DISABLE_CONNECTIVITY_LOCATION_CODE` preprocessor to make it operate without location permission. README documents a) remind users of `network_info`, b) suggest pinning to v0.5 as temporary workaround and c) guarantee to update library as needed to avoid requiring location permissions. This is the breaking change and where the pain happens.
5) v1.0: Some time later once the API has proved stable over time, upgrade major version at your leisure.

**Comment by brunobowden** [comment]
FYI @cyanglaz @KevinTheGray

**Breaking Change Disruption Analysis**

I've done an analysis of `getWifiName` and `getWifiBSSID` and it appears to be mostly copies of the entire Flutters/plugins repo... or mostly small adaptations of the example dart files. It's hard to be definitive on how much disruption there would be for breaking backwards compatibility but I think it will be modest.

The Flutter Plugins repo has 5 files names `test_connectivity.dart`. There's 116 files that match the same search across all of GitHub. Manually sampling a number of files, I couldn't see any noticeable modifications from the original source. That suggests a lot of copying of the Flutter/plugins repo that are not forks. This suggests ~23 copies of the Flutter/plugins repo (inc. the original one).

5 - [Flutter/plugins](https://github.com/flutter/plugins/search?q=extension%3Adart+filename%3Atest_connectivity.dart)
116 - [GitHub](https://github.com/search?q=extension%3Adart+filename%3Aconnectivity_test.dart)

Now searching for `getWifiName`, that gets 15 files matching in the Flutter/plugins repo. Given the ~23 copies of the repo, that means an expected ~345 copied repo files matching across all of GitHub. This compares to the 420 files found. Suggesting ~75 files that aren't whole copies of the repo.

420 - [GitHub: getWifiName](https://github.com/search?q=extension%3Adart+getWifiName)

Using the big assumption that files with the same name as in the repo haven't been modified much, the remaining files are mostly small adaptations of https://github.com/flutter/plugins/blob/master/packages/connectivity/connectivity_macos/example/lib/main.dart:

162 - [extension:dart getWifiName NOT public_member_api_docs -filename:test.dart -filename:connectivity.dart](https://github.com/search?p=4&q=extension%3Adart+getWifiName+NOT+public_member_api_docs+-filename%3Atest.dart+-filename%3Aconnectivity.dart&type=Code)
Example adaptation: https://github.com/Samdroid27/connectum/blob/eecee769d297686a2b82a8e6a7572fd8ac1fb558/lib/main.dart

Therefore, I believe the disruption would be irritating to a small number of users but far from unmanageable. It is hard to be definitive though on precise numbers.

**Comment by cyanglaz** [comment]
Fixed in connectivity 2.0
